// @bokeh/bokehjs/build/js/lib/models/tickers/timedelta_ticker.js
import { AdaptiveTicker } from "./adaptive_ticker";
import { CompositeTicker } from "./composite_ticker";
import { ONE_MILLI, ONE_SECOND, ONE_MINUTE, ONE_HOUR, ONE_DAY } from "./util";
export class TimedeltaTicker extends CompositeTicker {
    static __name__ = "TimedeltaTicker";
    constructor(attrs) {
        super(attrs);
    }
    static {
        this.override({
            num_minor_ticks: 0,
            tickers: () => [
                // Sub-second.
                new AdaptiveTicker({
                    mantissas: [1, 2, 5],
                    base: 10,
                    min_interval: 0,
                    max_interval: 500 * ONE_MILLI,
                    num_minor_ticks: 0,
                }),
                // Seconds, minutes.
                new AdaptiveTicker({
                    mantissas: [1, 2, 5, 10, 15, 20, 30],
                    base: 60,
                    min_interval: ONE_SECOND,
                    max_interval: 30 * ONE_MINUTE,
                    num_minor_ticks: 0,
                }),
                // Hours.
                new AdaptiveTicker({
                    mantissas: [1, 2, 4, 6, 8, 12],
                    base: 24.0,
                    min_interval: ONE_HOUR,
                    max_interval: 12 * ONE_HOUR,
                    num_minor_ticks: 0,
                }),
                // Days.
                new AdaptiveTicker({
                    mantissas: [1, 2, 5],
                    base: 10,
                    min_interval: ONE_DAY,
                    max_interval: null,
                    num_minor_ticks: 0,
                }),
            ],
        });
    }
}
