// @bokeh/bokehjs/build/js/lib/models/annotations/base_bar.js
import { Annotation, AnnotationView } from "./annotation";
import { LabelOverrides } from "../axes/axis";
import { FixedTicker } from "../tickers/fixed_ticker";
import { TickFormatter } from "../formatters/tick_formatter";
import { LabelingPolicy, NoOverlap } from "../policies/labeling";
import { BaseText } from "../text/base_text";
import { Anchor, Orientation } from "../../core/enums";
import * as mixins from "../../core/property_mixins";
export class BaseBarView extends AnnotationView {
    static __name__ = "BaseBarView";
}
export class BaseBar extends Annotation {
    static __name__ = "BaseBar";
    constructor(attrs) {
        super(attrs);
    }
    static {
        this.mixins([
            ["title_", mixins.Text],
            ["major_label_", mixins.Text],
            ["major_tick_", mixins.Line],
            ["minor_tick_", mixins.Line],
            ["background_", mixins.Fill],
            ["background_", mixins.Hatch],
            ["border_", mixins.Line],
            ["bar_", mixins.Line],
        ]);
        this.define(({ Float, Str, Tuple, Or, Ref, Enum, Auto, Nullable }) => ({
            location: [Or(Anchor, Tuple(Float, Float)), "top_right"],
            orientation: [Or(Orientation, Auto), "auto"],
            width: [Or(Float, Enum("max")), 200],
            height: [Or(Float, Enum("max")), 50],
            margin: [Float, 30],
            padding: [Float, 10],
            title: [Nullable(Or(Str, Ref(BaseText))), null],
            title_standoff: [Float, 2],
            ticker: [Or(Ref(FixedTicker), Auto), "auto"],
            formatter: [Or(Ref(TickFormatter), Auto), "auto"],
            major_label_overrides: [LabelOverrides, new Map()],
            major_label_policy: [Ref(LabelingPolicy), () => new NoOverlap()],
            label_standoff: [Float, 5],
            major_tick_in: [Float, 5],
            major_tick_out: [Float, 0],
            minor_tick_in: [Float, 0],
            minor_tick_out: [Float, 0],
        }));
        this.override({
            background_fill_color: "#ffffff",
            background_fill_alpha: 0.95,
            border_line_color: null,
            bar_line_color: null,
            major_label_text_font_size: "11px",
            major_tick_line_color: "black",
            minor_tick_line_color: null,
            title_text_font_size: "13px",
            title_text_font_style: "italic",
        });
    }
}
